docs/modules/SDDs/pages/0031-component-version-tracking.adoc

```diff
@@ -0,0 +1,143 @@
+Can be a branch name, tag, or commit hash.
+The `version` can point to different commits depending on when the compile was done.
+* `gitSha`: The commit hash of the commit referenced by the `version` field
+With the `gitSha` and the `url` it is possible to uniquely identify the commit.
+
+[source,yaml]
+----
+apiVersion: syn.tools/v1alpha1
+kind: Cluster
+metadata:
+  name: my-cluster
+status:
+  compileMeta:
+    lastCompile: "2024-05-13T12:00:00Z"
+    packages:
+      appcat:
+        url: https://github.com/vshn/component-appcat
+        version: master
+        gitSha: "125a78afd733f47d50d3eee9cd7c5bcaca2282b2"
+    components:
+      alerts-exporter:
+        url: https://github.com/appuio/component-alerts-exporter.git
+        version: "master"
+        gitSha: "3244ddb83a91279fb378f011358ac118987747a2"
+      rook-ceph:
+        url: https://github.com/projectsyn/component-rook-ceph.git
+        version: "v8.0.1"
+        gitSha: "3a3d940ff53d73307f4a86955af2a8e1a0b7961f"
+    global:
+      url: git.syn.tools/my-org/global-config.git
+      version: "master"
+      gitSha: "a64d207a11bb595702470917beffed9df0227a36"
+    tenant:
+      url: git.syn.tools/chewserver/syn-tenant-repo.git
+      version: "master"
+      gitSha: "4e6699968153c608f048b140f6af351816b14303"
+----
+
+=== API
+
+The Lieutenant API will be extended to have a push endpoint `POST /clusters/{clusterId}/compileMeta` with the same token authentication as the current API.
+The API accepts a JSON payload with the contents of the `compileMeta` field and handles the update of the `Cluster` CRD.
+
+=== Operator
+
+The Lieutenant operator will be extended to create Prometheus / OpenMetrics metrics for the component versions.
+The metrics should contain all the information from the `compileMeta` field.
+
+[source]
+----
+syn_lieutenant_cluster_compile_meta_last_compile{cluster="my-cluster"} 1.624e+12
+syn_lieutenant_cluster_compile_meta_component{cluster="my-cluster", component="alerts-exporter", url="https://...", version="master", gitSha="..."} 1
+syn_lieutenant_cluster_compile_meta_component{cluster="my-cluster", component="rook-ceph", url="https://...", version="v8.0.1", gitSha="..."} 1
+syn_lieutenant_cluster_compile_meta_global{cluster="my-cluster", url="https://...", version="master", gitSha="..."} 1
+syn_lieutenant_cluster_compile_meta_tenant{cluster="my-cluster", url="https://...", version="master", gitSha="..."} 1
+----
+
+=== Implementation Details/Notes/Constraints
+
+Different versions for different component instances::
+https://github.com/projectsyn/commodore/issues/563[projectsyn/commodore#563] wants to allow different versions for different component instances.
+Since the component name is currently the key to the version information, this can't be supported.
+One idea is to call the field `instances` instead of `components`, use the instance name as the key, and add the component name as a field in the value.
```

Comment:
We could already go with the proposed structure that uses top-level key `instances` and already implement the reporting with this format in anticipation of implementing this feature at some point (technically all components are already treated as "instances" by Commodore internally, most of the time we just use the "identity instance" which has the same name as the component itself):

```
instances
  keycloak:
    component: keycloak
    url: https://github.com/projectsyn/component-keycloak.git
    version: v1.2.3
    gitSha: ...
  keycloak-prod:
    component: keycloak
    url: https://github.com/projectsyn/component-keycloak.git
    version: v1.2.3
    gitSha: ...
```

Side-note: this structure would probably work reasonably well for defining versions, by making the appropriate fields optional: make `component` optional and only require `url` and `version` if `component` is absent; inherit `url` and `version` from the entry which has the key given in `component` when `component` is present; allow url & version overrides from the current entry; extend the linter to disallow loops, for example by making sure that entries which have field `component` never appear as value in field `component` since we don't need deeply nested configs anyway.

Reply:
follow-up: cf. the note in https://syn.tools/commodore/reference/architecture.html#_component_instantiation for details on the internal handling of unaliased components

Reply:
added to the proposal!